lahja/common.py

```diff
@@ -0,0 +1,111 @@
+from abc import (
+    ABC,
+    abstractmethod,
+)
+from typing import (
+    TYPE_CHECKING,
+    Generic,
+    NamedTuple,
+    Optional,
+    Set,
+    Type,
+    TypeVar,
+    Union,
+)
+
+from lahja.misc import (
+    BroadcastConfig,
+)
+
+if TYPE_CHECKING:
+    from lahja.base import EndpointAPI
+
+
+class BaseEvent:
+    __slots__ = ("_origin", "_id")
+
+    @property
+    def is_bound(self) -> bool:
+        return hasattr(self, '_origin')
+
+    def bind(self,
+             endpoint: 'EndpointAPI',
+             id: Optional[str] = None) -> None:
```

Comment:
@cburgdorf curious for feedback on this API.  I wanted to normalize the lifecycle of events, specifically setting of `_id` and `_origin`.  You can see some of this being used in the mechanisms for establishing reverse connections in the `_WhoAreYou` and `_MyNameIs` events.

Reply:
Makes sense. It always looked a bit odd that we were setting private properties.